fix: refactor LightweightUserAdapter instantiation for DefaultTokenEx…

nbaiocco opened the pull request:
Avoid NullPointerException during external to internal token-exchange iff transient user is enabled. Modified LightweightUserAdapter instantiation to include the realm parameter to avoid NullPointerException in org.keycloak.storage.adapter.AbstractInMemoryUserAdapter#getGroupsStream.

Closes #28241

nbaiocco committed:
fix: refactor LightweightUserAdapter instantiation for DefaultTokenExchangeProvider

Avoid NullPointerException during external to internal token-exchange iff transient user is enabled.
Modified LightweightUserAdapter instantiation to include the realm parameter to avoid NullPointerException in org.keycloak.storage.adapter.AbstractInMemoryUserAdapter#getGroupsStream.

Closes #28241

Signed-off-by: Nicola Baiocco <nicola.baiocco@intesys.it>

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/DefaultTokenExchangeProvider.java:
```diff
@@ -630,7 +630,7 @@ protected UserModel importUserFromExternalIdentity(BrokeredIdentityContext conte
 
             if (context.getIdpConfig().isTransientUsers()) {
-                user = new LightweightUserAdapter(session, context.getAuthenticationSession().getParentSession().getId());
+                user = new LightweightUserAdapter(session, realm);
```
```suggestion
                user = new LightweightUserAdapter(session, realm, context.getAuthenticationSession().getParentSession().getId());
```

Is it possible to use that one above instead of generated ID? Or any particular reason why ID needs to be changed to be generated?

If you change this way, you can also remove constructor with signature `public LightweightUserAdapter(KeycloakSession session, RealmModel realm)` as it would not used.

nbaiocco replied:
As mentioned in #28241, the NullPointerException is caused by context.getAuthenticationSession() == null, for this reason, I removed that part. 
If you believe this removal is incorrect, I can investigate further to understand why context.getAuthenticationSession() of BrokeredIdentityContext is null.

In my tests, both context.getAuthenticationSession() and context.getBrokerSessionId() (of BrokeredIdentityContext) consistently return null.

nbaiocco replied:
Ok, I applied your suggestions in the second commit. 
It is slightly different, but it ensures the NullPointerException is avoided in cases where the parent session is empty.

mposolda reviewed (commented):
@nbaiocco Thanks for the PR, one small suggestion added

nbaiocco committed:
fix: pass authSessionId to LightweightUserAdapter

Enable passing of authentication session ID to LightweightUserAdapter for transient users in DefaultTokenExchangeProvider. This improves session tracking and maintains user context. Removed an unused constructor in LightweightUserAdapter for cleaner code.

Signed-off-by: Nicola Baiocco <nicola.baiocco@intesys.it>

mposolda reviewed (approved):
@nbaiocco Thanks!
